crates/lib/src/qpu/runner.rs

```diff
@@ -9,132 +6,128 @@ use num::complex::Complex32;
+
+pub(crate) fn params_into_job_execution_configuration(
+    params: &Parameters,
+) -> JobExecutionConfiguration {
+    let memory_values = params
+        .iter()
+        .map(|(str, value)| {
+            (
+                str.as_ref().into(),
+                DataValue {
+                    value: Some(Value::Real(RealDataValue {
+                        data: value.to_vec(),
+                    })),
+                },
+            )
+        })
+        .collect();
+
+    JobExecutionConfiguration { memory_values }
+}
 
 /// The QCS Job ID. Useful for debugging or retrieving results later.
 #[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
 pub struct JobId(pub(crate) String);
 
 /// Execute compiled program on a QPU.
-pub(crate) fn submit(
-    program: &str,
+pub(crate) async fn submit(
+    quantum_processor_id: &str,
+    program: EncryptedControllerJob,
     patch_values: &Parameters,
-    client: &Client,
-) -> Result<JobId, Error> {
-    let params = QPUParams::from_program(program, patch_values);
-    let request = RPCRequest::from(&params);
-    let job_id: String = client.run_request(&request)?;
-    Ok(JobId(job_id))
+    client: &QcsClient,
+) -> Result<JobId, ClientGrpcError> {
+    let request = ExecuteControllerJobRequest {
+        execution_configurations: vec![params_into_job_execution_configuration(patch_values)],
+        job: Some(execute_controller_job_request::Job::Encrypted(program)),
+        target: Some(execute_controller_job_request::Target::QuantumProcessorId(
+            quantum_processor_id.into(),
+        )),
+    };
+
+    // we expect exactly one job ID since we only submit one execution configuration
+    let job_execution_id = client
+        .get_controller_client(quantum_processor_id)
+        .await?
+        .execute_controller_job(request)
+        .await?
+        .into_inner()
+        .job_execution_ids
+        .pop();
+
+    job_execution_id
+        .map(JobId)
+        .ok_or_else(|| ClientGrpcError::ResponseEmpty("Job Execution ID".into()))
 }
 
-pub(crate) fn retrieve_results(
+pub(crate) async fn retrieve_results(
     job_id: JobId,
-    client: &Client,
-) -> Result<GetExecutionResultsResponse, Error> {
-    let get_buffers_request = GetExecutionResultsRequest::new(job_id.0);
+    quantum_processor_id: &str,
+    client: &QcsClient,
+) -> Result<ControllerJobExecutionResult, ClientGrpcError> {
+    let request = GetControllerJobResultsRequest {
+        job_execution_id: Some(job_id.0),
+        target: Some(
+            get_controller_job_results_request::Target::QuantumProcessorId(
+                quantum_processor_id.into(),
+            ),
+        ),
+    };
+
     client
-        .run_request(&RPCRequest::from(&get_buffers_request))
-        .map_err(Error::from)
+        .get_controller_client(quantum_processor_id)
+        .await?
+        .get_controller_job_results(request)
+        .await?
+        .get_ref()
+        .result
+        .to_owned()
```

Comment:
This seems redundant. Is there not a way to get the owned value directly?

Reply:
Yeah, I think we can do `get_controller_job_results(request).await?.into_inner()`